py/server/deephaven/table_listener.py

```diff
@@ -442,3 +445,173 @@ def listen(t: Table, listener: Union[Callable, TableListener], description: str
+
+        self.j_listener_recorder = _JListenerRecorder("Python Wrapped Listener recorder", table.j_table, None)
+        self.table = table
+
+    def are_recorded_updates_valid(self) -> bool:
+        """Checks if the recorded updates are valid for the current update cycle.
+
+        Returns:
+            a bool
+        """
+        return self.j_listener_recorder.recordedVariablesAreValid()
+
+    def table_update(self) -> Optional[TableUpdate]:
+        """Gets the table update from the listener recorder. If there is no update in the current update graph cycle,
+        returns None.
+
+        Returns:
+            a TableUpdate or None
+        """
+        j_table_update = self.j_listener_recorder.getUpdate()
+        return TableUpdate(self.table, j_table_update) if j_table_update else None
+
+
+class MergedListener(ABC):
+    """A MergedListener has a collection of ListenerRecorder. Each one must complete before the merged listener
+    executes only once for the current update graph cycle. """
+
+    listener_recorders: Sequence[ListenerRecorder]
+
+    @abstractmethod
+    def process(self) -> None:
+        """The required method on a user defined subclass of MergedListener that processes the table updates from the
+        list of ListenerRecorders automatically provided after being adapted by the MergedListenerHandle.
+        The implementation of this method in the subclass can reference the listener recorders by the
+        'listener_recorders' attribute. The listener recorders are in the same order as the source tables provided during
+        the instance initialization.
+
+        Example:
+            def process(self) -> None:
+                for i, listener in enumerate(self.listener_recorders):
+                    print(f"table {i}:", self.listener_recorders[i].table_update())
+        """
+        ...
+
+
+class MergedListenerHandle:
+    """A handle to manage a merged listener's lifecycle."""
+
+    def __init__(self, listener_recorders: Sequence[ListenerRecorder], listener: Union[Callable, MergedListener],
+                 description: str = None, dependencies: Union[Table, Sequence[Table]] = None):
+        """Creates a new MergedListenerHandle with the provided listener recorders and dependencies.
+
+        Table change events are processed by 'listener', which can be either
+        (1) a callable (e.g. function) or
+        (2) an instance of MergedTableListener type which provides a "process" method.
+        The callable or the process method must have zero parameters.
+
+        Note: Don't do table operations in the listener. Do them beforehand, and add the results as dependencies.
+
+        Args:
+            listener_recorders (Sequence[ListenerRecorder]): listener recorders that record the table updates for the
+                merged listener to process.
+            listener (Union[Callable, MergedListener]): listener to process table updates from the listener recorders
```

Comment:
Why do we offer `MergedListener` as an option? I guess what I'm wondering is if we need this flexibility. Maybe we want to offer the user a callback that takes a dict of `{Table, ListenerRecorder}` or even `{Table, TableUpdate}`, instead of one that just calls `process()`?

Reply:
The new `MergedListener` and the existing `TableListener` interfaces are offered to support use cases when the listeners need to maintain some state across multiple calls. Strictly speaking they are not necessary but can be useful when  template for the users.

I struggled with the design of the interface quite a bit. I am definitely for making the interface more user friendly but I don't want to lose the flexibility and extensibility  that comes with exposing `ListenerRecorder`.  I ended up with this interface because I favor explicitness and simplicity, yet I am not very happy with the implicitness that comes with the no-arg listeners.

My thoughts on
1. a listener which takes an argument of  `Dict[Table, ListenerRecorder]` is good,  though we do still need to explain where the `[ListenerRecorder]` comes from.
2. a listener which takes an argument of `Dict[Table, TableUpdate]` is easier to use for sure, but in case we need to expose more functionality from `ListenerRecorder` we'll have more work to do to make it happen.

So of the two alternatives I favor 1.

Reply:
@chipkent Could you provide some feedback?